### ISSUE TITLE
Support for metrics in OpenTelemetry.Instrumentation.Hangfire

### ISSUE THREAD

**Issue by Toumash** [open-telemetry/opentelemetry-dotnet-contrib#2075]
### Component

OpenTelemetry.Instrumentation.Hangfire

### Is your feature request related to a problem?

Currently to use hangfire metrics you have either:
1. Install separate exporter
2. Use prometheus directly with hangfire to prometheus exporter

### What is the expected behavior?

I want to install OpenTelemetry.Instrumentation.Hangfire and have all the metrics from nuget `Hangfire.Prometheus.NetCore`

### Which alternative solutions or features have you considered?

nuget `Hangfire.Prometheus.NetCore`

**Comment by gorbach** [comment]
Hi! I'd like to contribute metrics support for Hangfire instrumentation.

I have a working implementation that provides:
- `hangfire.job.duration` (histogram) - Job execution duration
- `hangfire.job.executions` (counter) - Total job executions
- `hangfire.job.active` (up/down counter) - Currently executing jobs
- `hangfire.job.queue.duration` (histogram) - Queue wait time

All metrics include rich tags: `job.type`, `job.method`, `job.status`, `exception.type`

**Before I start a PR, I'd like guidance on:**

1. **Metric naming conventions** - Should I follow specific semantic conventions for job/background task metrics?
2. **Integration approach** - Should this extend existing `HangfireInstrumentationOptions` or be separate?
3. **Testing requirements** - What test coverage is expected for metrics instrumentation?
4. **API surface** - Should metrics be opt-in/opt-out? Same pattern as tracing?

**Implementation approach I'm considering:**
- New `HangfireMetrics.cs` implementing `IServerFilter` (mirrors existing tracing filter)
- Extension method: `MeterProviderBuilder.AddHangfireInstrumentation()`
- Use `System.Diagnostics.Metrics` API (similar to ProcessMetrics pattern)
- Hook into `OnPerforming` and `OnPerformed` lifecycle events

Happy to discuss the design before starting the PR. I can also share my current implementation if that helps!

**Comment by Toumash** [comment]
Maybe lets share it as a draft PR, then if needed anyone can easily continue the process.
I cannot help with the guidance for now, as i simply dont know


For now i'm using `prometheus-net`
<details>

```csharp
using Hangfire.Common;
using Hangfire.Server;
using Microsoft.Extensions.Logging;

public sealed class JobDurationFilter(IJobDurationMetrics metrics, ILogger<JobDurationFilter> logger) : JobFilterAttribute, IServerFilter
{
    private const string T0Key = "job-t0";

    public void OnPerforming(PerformingContext context)
    {
        var (queue, type, method) = GetJobSignature(context);
        context.Items[T0Key] = System.Diagnostics.Stopwatch.GetTimestamp();
    }

    private static readonly double TickToSeconds = 1.0 / System.Diagnostics.Stopwatch.Frequency;

    public void OnPerformed(PerformedContext context)
    {
        var (queue, type, method) = GetJobSignature(context);

        var t0 = context.Items.TryGetValue(T0Key, out var v) && v is long ts ? ts : System.Diagnostics.Stopwatch.GetTimestamp();
        var t1 = System.Diagnostics.Stopwatch.GetTimestamp();
        var seconds = (t1 - t0) * TickToSeconds;
        bool success = context.Exception is null;

        metrics.ObserveDuration(queue, type, method, seconds, success);
        logger.LogInformation("Job timing: {Queue} {Type}.{Method} took {Seconds:F3}s (success={Success})",
            queue, type, method, seconds, success);
    }

    private static (string queue, string type, string method) GetJobSignature(PerformContext ctx)
    {
        var job = ctx.BackgroundJob?.Job;
        string queue = job?.Queue ?? "default";
        string type = job?.Type?.FullName ?? "unknown";
        string method = job?.Method?.Name ?? "unknown";
        return (queue, type, method);
    }
}
public interface IJobDurationMetrics
{
    // seconds is wall-clock duration
    void ObserveDuration(string queue, string type, string method, double seconds, bool success);
}

public sealed class PrometheusJobDurationSummaryMetrics : IJobDurationMetrics
{
    private static readonly Summary Duration = Prometheus.Metrics.CreateSummary(
        "hangfire_job_duration_seconds",
        "Hangfire job duration (quantiles over a sliding window 24h).",
        new SummaryConfiguration
        {
            LabelNames = new[] { "queue", "type", "method", "result" },
            // 10-min sliding window, ~1-min age buckets (prometheus-net defaults)
            // Customize if you want a different retention/resolution:
            MaxAge = TimeSpan.FromHours(24),
            AgeBuckets = 24,
            Objectives = new[]
            {
                new QuantileEpsilonPair(0.5, 0.01),
                new QuantileEpsilonPair(0.9, 0.01),
                new QuantileEpsilonPair(0.99, 0.005),
            }
        });


    public void ObserveDuration(string q, string t, string m, double s, bool ok) =>
        Duration.WithLabels(q, t, m, ok ? "success" : "fail").Observe(s);
}
public class JobLoggerAttribute : JobFilterAttribute, IClientFilter, IServerFilter
{
    private ILogger<JobLoggerAttribute> _logger;

    public JobLoggerAttribute(IServiceProvider serviceProvider)
    {
        _logger = serviceProvider.GetRequiredService<ILogger<JobLoggerAttribute>>();
    }

    public void OnCreating(CreatingContext filterContext)
    {
        _logger.LogInformation("Job is being created for {JobType} with arguments {JobArguments}",
            filterContext.Job.Type.Name, filterContext.Job.Args);
    }

    public void OnCreated(CreatedContext filterContext)
    {
        _logger.LogInformation("Job {JobId} has been created", filterContext.BackgroundJob.Id);
    }

    public void OnPerforming(PerformingContext filterContext)
    {
        _logger.LogInformation("Job {JobId} is performing", filterContext.BackgroundJob.Id);
    }

    public void OnPerformed(PerformedContext filterContext)
    {
        _logger.LogInformation("Job {JobId} has performed", filterContext.BackgroundJob.Id);
        if (filterContext.Exception != null)
        {
            _logger.LogError(
                filterContext.Exception,
                "Job {JobId} failed due to an exception",
                filterContext.BackgroundJob.Id);
        }
    }
}
public interface IJobMetrics
{
    // Records one execution outcome (success/fail) and tags OOMs.
    void ObserveExecution(string queue, string type, string method, bool success, bool isOom);
}

public sealed class PrometheusJobMetrics : IJobMetrics
{
    private static readonly Counter ExecCounter = Metrics.CreateCounter(
        "hangfire_job_executions_total",
        "Count of Hangfire job executions by outcome.",
        new CounterConfiguration
        {
            LabelNames = new[] { "queue", "type", "method", "result", "exception" }
        });

    private static readonly Counter OomCounter = Metrics.CreateCounter(
        "hangfire_job_oom_total",
        "Count of Hangfire jobs that ended with an OutOfMemoryException.",
        new CounterConfiguration
        {
            LabelNames = new[] { "queue", "type", "method" }
        });

    public void ObserveExecution(string queue, string type, string method, bool success, bool isOom)
    {
        // Keep exception label low-cardinality: "none" | "oom" | "other"
        var exception = success ? "none" : (isOom ? "oom" : "other");
        ExecCounter.WithLabels(queue, type, method, success ? "success" : "fail", exception).Inc();

        if (isOom)
            OomCounter.WithLabels(queue, type, method).Inc();
    }
}

```

</details>

**Comment by Kielek** [comment]
1. **Metric naming conventions** - Should I follow specific semantic conventions for job/background task metrics?
    I do not know exact answer here. You should ask for the advice under semantic convention repository: https://github.com/open-telemetry/semantic-conventions/issues.
2. **Integration approach** - Should this extend existing `HangfireInstrumentationOptions` or be separate?
    Separate, you can check couple stable packages. Typically it should be `*TraceInstrumentationOptions` + `*MetricsInstrumentationOptions`. I would say that rename deserves separate PR + changelog entry marked as `**Breaking change:** `
3. **Testing requirements** - What test coverage is expected for metrics instrumentation?
    I hope @fred2u can advice here. He is component owner and subject matter expert. 
6. **API surface** - Should metrics be opt-in/opt-out? Same pattern as tracing?
    Opt-in,  check other `*InstrumentationMeterProviderBuilderExtensions` classes.

@fred2u, FYI as you are the only component owner of this package.

**Comment by Kielek** [comment]
One more thing, if you decide to cleanup name for TracerInstrumentationOption, please rename also `TracerProviderBuilderExtensions`. It should be prefixed by the name of instrumentation ---> `HangfireTracerProviderBuilderExtensions`. With this, there should be no conflicts while you reference more the one instrumentation package in the place where you want to build whole OTel infrastructure.

**Comment by mr-miles** [comment]
I've just started using this package, and ended up here because I was keen to get metrics and was going to make a proposal.

Happy to help with testing or whatever is useful as metrics would be really helpful.
